<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <title>Mrcchip by ThomasCarroll</title>

    <link rel="stylesheet" href="stylesheets/styles.css">
    <link rel="stylesheet" href="stylesheets/pygment_trac.css">
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
  </head>
  <body>
    <div class="wrapper">
      <header>
        <h1>Mrcchip</h1>
        <p>A pipeline for chipseq</p>

        <p class="view"><a href="https://github.com/ThomasCarroll/mrcchip">View the Project on GitHub <small>ThomasCarroll/mrcchip</small></a></p>


        <ul>
          <li><a href="https://github.com/ThomasCarroll/mrcchip/archive/chippipelineV0.1.zip">Download <strong>ZIP File</strong></a></li>
          <li><a href="https://github.com/ThomasCarroll/mrcchip/archive/chippipelineV0.1.tar.gz">Download <strong>TAR Ball</strong></a></li>
          <li><a href="https://github.com/ThomasCarroll/mrcchip">View On <strong>GitHub</strong></a></li>
        </ul>
      </header>
      <section>
        <h1>
<a name="chip-seq-qc-pipeline" class="anchor" href="#chip-seq-qc-pipeline"><span class="octicon octicon-link"></span></a>ChIP seq QC pipeline</h1>

<p>The chipseq pipeline provides basic processing, analysis and quality control applicable to ChIP-seq and other targeted enrichment sequencing technologies. The full description of these steps can be seen below.
Currently this is designed for use on the Imperial ax3 cluster using pbs pro v 12.2:</p>

<h5>
<a name="to-get-a-working-copy-of-this-project-please-use-the-version-installed-on-the-ax3-contact-thomascarrollimperialacuk" class="anchor" href="#to-get-a-working-copy-of-this-project-please-use-the-version-installed-on-the-ax3-contact-thomascarrollimperialacuk"><span class="octicon octicon-link"></span></a>To get a working copy of this project please use the version installed on the ax3. (Contact <a href="mailto:thomas.carroll@imperial.ac.uk">thomas.carroll@imperial.ac.uk</a>).</h5>

<p>If you would like the review the source code download from this <a href="https://github.com/ThomasCarroll/mrcchip/archive/0.1.zip">link</a>. </p>

<pre><code>
$ wget https://github.com/ThomasCarroll/mrcchip/archive/chippipelineV0.1.zip ./
$ unzip chippipelineV0.1.zip
</code></pre>

<h2>
<a name="quick-start-guide" class="anchor" href="#quick-start-guide"><span class="octicon octicon-link"></span></a>Quick start guide</h2>

<p>** Coming soon**</p>

<h2>
<a name="running-the-pipeline" class="anchor" href="#running-the-pipeline"><span class="octicon octicon-link"></span></a>Running the pipeline</h2>

<h4>
<a name="configuration" class="anchor" href="#configuration"><span class="octicon octicon-link"></span></a>Configuration</h4>

<p>This section covers the basics configuration steps for running the chipseq pipeline including:</p>

<ul>
<li>Working directories</li>
<li>species</li>
<li>technical replicate merging.</li>
<li>Input requirements and peak calling parameters.</li>
<li>Sample and replicate groupings.</li>
<li>Reporting requirements</li>
</ul><p>The species to align to is provided as a command line argument, with the directory to output results {see below} where as all other parameters are provided in the samplesheet (<a href="https://github.com/ThomasCarroll/mrcchip/blob/master/example_samplesheet.txt">example samplesheet</a>)</p>

<pre><code>usage: ./runAln.py -S(--samplesheet) [path_to_samplesheet] -D(--directory) [path_to_directory] -G(--genome) [genome]

</code></pre>

<h3>
<a name="working-directories" class="anchor" href="#working-directories"><span class="octicon octicon-link"></span></a>Working Directories</h3>

<p>The working directory argument controls where the base directory for the pipeline output is. Fro there the pipeline will create the various job output and error directories and all subsequent processed and temporary files will be directed here.</p>

<p>The basic structure of the output/error directories is.</p>

<ul>
<li>
<strong>AlignedData</strong> - <em>Aligned sequence data in bam format</em>
</li>
<li>
<strong>Coverage</strong> - <em>Raw and normalised signal graph files in bigwig format</em>
</li>
<li>
<strong>flagstat</strong> - <em>Samtools flagstat output for aligned sequence file</em>
</li>
<li>
<strong>Macs</strong> - <em>Macs 2 peak calls</em>
</li>
<li>
<strong>chipqc</strong> - <em>R data files of ChIPQCsample objects containing per sample quality.</em>
</li>
</ul><p>A final project *.RData file can also be found at this level. This contains summary information for whole project/experiment/pipeline run.</p>

<ul>
<li>
<strong>[samplesheet_name]report.RData</strong> - ChIPQCexperiment object summarising project ChIP quality.</li>
</ul><h3>
<a name="species" class="anchor" href="#species"><span class="octicon octicon-link"></span></a>Species</h3>

<p>Currently this pipeline supports 3 genomes (<strong>hg19</strong>,<strong>mm9</strong>,<strong>dm3</strong>) and all genomes were retrieved from illumina's <a href="http://support.illumina.com/sequencing/sequencing_software/igenome.ilmn">IGenomes</a>:</p>

<ul>
<li>Homo sapiens (Ensembl)  <strong>hg19</strong>
</li>
<li>Mus musculus  (UCSC) <strong>mm9</strong>
</li>
<li>Drosophila melanogaster  (UCSC)  <strong>dm3</strong>
</li>
</ul>
      </section>
      <footer>
        <p>This project is maintained by <a href="https://github.com/ThomasCarroll">ThomasCarroll</a></p>
        <p><small>Hosted on GitHub Pages &mdash; Theme by <a href="https://github.com/orderedlist">orderedlist</a></small></p>
      </footer>
    </div>
    <script src="javascripts/scale.fix.js"></script>
    
  </body>
</html>
